### DISCUSSION TITLE
articles/2024/django-vs-flask-vs-fastapi/

### DISCUSSION THREAD

**Discussion by giscus** [loopwerk/loopwerk.io#22] [Article discussions]
# articles/2024/django-vs-flask-vs-fastapi/

I started using Django in 2009, and fifteen years later I am still a happy user. Why do I prefer it over Flask or FastAPI?

https://www.loopwerk.io/articles/2024/django-vs-flask-vs-fastapi/

**Comment by fcovatti** [comment]
There is one important note that you forgot to mention: Usually the bottleneck of performance is at the Database level. So no matter if you use FastAPI or Django, the query performance will be the most critical point impacting the response times of your application. So a cache mechanism, some adjustment in object serialization can improve performance by a lot. So no matter if you choose one or the other, if you improve the query performance you can even have faster performance on API level on Django then on FastAPI.

**Reply by kevinrenskers** [reply]
Very true! But for me the performance is just not an issue with Django - as long as you keep an eye on the queries, that is. It's why I value the other things so much more than the pure performance improvements that FastAPI _could_ potentially bring.

**Reply by lukasvinclav** [reply]
Exactly. Most of the time the issue is just simple N+1 SQL queries which can be fixed by using `prefetch_related` or `select_related`. In all of my projects, I'm always using django-debug-toolbar to track these problems.

**Comment by lukasvinclav** [comment]
Nice list! For me, the ORM and migrations are the best parts of Django. I would add just one more thing to the list: the smoothness of Django's upgradability. You don’t have to refactor the application just because there’s a new version, and most things just work.

Regarding the admin part, I know what you mean, and that’s the reason I created the Unfold admin theme. You can take a look at it here: https://github.com/unfoldadmin/django-unfold.

**Reply by kevinrenskers** [reply]
You're absolutely right. Although on the other hand.. I do wish that Django would get rid of some old cruft and break backwards compatibility for once, for the greater good 🙈

Nice looking theme! I'll try to find some time to play around with it, although usually I am a bit scared of using third party themes. Relying on them to keep it working with new releases etc.

**Comment by Jdsleppy** [comment]
Well said.  One correction, FastAPI can make a server-rendered HTML app very well!  It is based on Starlette which gives it some batteries included: Jinja templating, a function like "reverse" for URLs, sessions, cookies, minimal authentication.

**Reply by kevinrenskers** [reply]
Are you using FastAPI at that point though? :)

**Reply by Jdsleppy** [reply]
I was writing an API in FastAPI and the UI in Starlette once, then I migrated the UI to FastAPI for reasons I forget and it was much nicer to work with.  So yes I think so!  Also using FastAPI makes it easy to have a little API in the same codebase.

**Comment by agouliel** [comment]
I love both Django and Flask. (Haven't tried FastAPI). Django was my first Python web framework, and I have to admit that the selling point for me was the Admin. Since then, I have built many projects and APIs using Django. But Flask is great as well. I was drawn into Flask by the excellent writings of Miguel Grinberg, and it remains my best choice when I want to quickly transform database queries into web endpoints.
